### PR TITLE
Future builder enhancements

### PR DESCRIPTION
**Pull request by pstromberg98**
This PR adds three parameters to the FutureBuilder.

`listen`: Adds the ability to trigger side effects outside of the build cycle whenever the future state changes
`onPending`: Widget builder that is called when the future is pending
`onError`: Widget builder that is called when the future throws

Refer to issue https://github.com/flutter/flutter/issues/87399.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [X] All existing and new tests are passing.

### CODE REVIEW

**Comment by cedvdb** [comment]
so the `builder` is only called when there is data ?

**Comment by pstromberg98** [comment]
@cedvdb If you use `onPending` and `onError`. It would be appropriate to also have `onComplete` there as well

**Comment by cedvdb** [comment]
Yes I agree but then it renders the original builder useless. Which honestly I find more readable.

**Comment by pstromberg98** [comment]
@cedvdb It renders the original builder useless only if you use the `onX` methods. So you would be able to to use the FutureBuilder the way you preferred. Why do you find the builder more readable? I feel like `onComplete`, `onPending`, and `onError` are much easier to understand and read than various if statements comparing conditions on an `AsyncSnapshot`.

I agree that the Flutter team most likely would want to keep the builder method for various reasons, one primary reason is to be consistent with builder widget types. That is why it is not deprecated in the PR

**Comment by cedvdb** [comment]
>  I feel like onComplete, onPending, and onError are much easier to understand and read than various if statements comparing conditions on an AsyncSnapshot

That's what I meant

File: packages/flutter/lib/src/widgets/async.dart [hunk 1]
```diff
@@ -737,6 +743,17 @@ class FutureBuilder<T> extends StatefulWidget {
   /// [AsyncSnapshot.hasError] will be true.)
   final T? initialData;
 
+  /// The builder used when the [AsyncSnapshot] is in the [ConnectionState.waiting] state.
+  final WidgetBuilder? onPending;
+
+  /// The builder used when the [AsyncSnapshot] has an error.
+  final WidgetBuilder? onError;
+
+  /// The listener that is called when the passed in [Future] state changes.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/async.dart:752]: Why does this have to be part of the widget? You can just listen to the future directly?

File: packages/flutter/lib/src/widgets/async.dart [hunk 2]
```diff
@@ -737,6 +743,17 @@ class FutureBuilder<T> extends StatefulWidget {
   /// [AsyncSnapshot.hasError] will be true.)
   final T? initialData;
 
+  /// The builder used when the [AsyncSnapshot] is in the [ConnectionState.waiting] state.
+  final WidgetBuilder? onPending;
+
+  /// The builder used when the [AsyncSnapshot] has an error.
+  final WidgetBuilder? onError;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/async.dart:750]: I am not exactly sure what this is enabling people to do that they cannot do today? It seems like it is just making the API more complex?

**Review by goderbauer — COMMENTED** [review]
This PR is also missing tests.

File: packages/flutter/lib/src/widgets/async.dart (hunk 2 above)
Comment (cedvdb) [reply packages/flutter/lib/src/widgets/async.dart:750]: it's less complex if there weren't two APIs, now it's just two API put together which is confusing

**Comment by ali2236** [comment]
I don't think `FutureBuilder` needs this. while I don't like using Snapshots and ConnectionState I don't think changing FutureBuilders API is a good idea and having multiple ways of doing the same thing is overall a bad practice. maybe instead we need another more intuitive FutureBuilder widget.

**Comment by pstromberg98** [comment]
@goderbauer Thanks for your review comments! The link above should give you some context about this PR. I also want to note that this PR is intended as a rough draft and as such does not have tests (one of the reasons I marked this as a draft PR).

@luca-venir Brought up two issues they were having with `FutureBuilder`. Which can be summed up as follows

1st: It is not straight forward on where to put business logic when using `FutureBuilder`. While it is possible to listen to the future outside of the build method `initState` for example, it is argued that a `listen` method on the `FutureBuilder` should be added as an easy way to add non-ui specific logic. This is the reasoning behind `listen`

2nd: The builder and `AsyncSnapshot` combo is confusing when first approaching `FutureBuilder`. The expected interface would be one that is more parallel to the possible states of a `Future` instead of a generic asynchronous representation (`AsyncSnapshot`). This is the reasoning behind `onError`, `onComplete`, and `onPending`.

I'd like to point out that while I have some opinions on the changes, I am not the originator of the ideas above. I was more interested in implementing them and possibly jumped the gun by creating a draft PR.

File: packages/flutter/lib/src/widgets/async.dart (hunk 2 above)
Comment (pstromberg98) [reply packages/flutter/lib/src/widgets/async.dart:750]: Potentially. The full rework from https://github.com/flutter/flutter/issues/87399 proposed breaking changes which this PR avoids.

**Comment by pstromberg98** [comment]
@goderbauer I worry that too many people are assuming that this PR is a source of truth and the intended implementation of https://github.com/flutter/flutter/issues/87399. It was merely a prospective first attempt and a place to discuss the specific implementation, thus the "draft" status. I think discussion on API design and use cases would be more appropriate at the https://github.com/flutter/flutter/issues/87399 thread. Unless you have any argument otherwise I will close this for the time being

**Comment by cedvdb** [comment]
listening to a future in the initState is inconvenient because you need a stateful widget.

**Comment by luca-venir** [comment]
> listening to a future in the initState is inconvenient because you need a stateful widget and disposing of said future, while if you already have a future builder that logic is handled for you

Hello there,

this is another problem in `FutureBuilder`, I guess the best place to discuss this is #87399. Unluckily, when a Future has to be dealt with (i.e. fired) into a `Provider` state-management solution, the Widget has to be stateful from my understanding. Otherwise, if the future is fired inside the build method, the async call will be fired at every re-build.

**Pull request closed without merging**